### ISSUE TITLE
Gradient reversal layer 

### ISSUE THREAD

**Issue by bbdamodaran** [VisionLearningGroup/SSDA_MME#4]
Hi, 
thanks for publishing the codes.
I have a question regarding the training with the gradient reversal layer.
As I understood from the code (main.py), Training consists of two steps:
1)  you update the feature extractor and classifier with respect to label data
2) you update the feature extractor and classifier with respect to unlabeled data using gradient reversal layer.

My question is whether you need two steps because in the paper you mentioned it is done one step.
whether it can't be done like this:
 - loss.backward() with labeled data
 - t_loss.backward() with unlabeled data
 - update the model  - optimizer.step()

Thanks in advance

**Comment by ksaito-ut** [comment]
Hi, thank you for your post and for pointing out this issue, which we did not notice. 
As long as I remember, what we tried to argue in the paper is the training step for unlabeled examples can be done with one step by using a gradient reversal layer. 
But, we did not implement the current two-steps-update training intentionally, and I think our method works with the one-step-update as you mention. 
We used the current two-steps update implementation to get the results of our paper, so we will keep this form. But, I will add some notes with regard to the point. Thanks so much for your valuable comments!

**Comment by bbdamodaran** [comment]
Thanks, @ksaito-ut for your prompt response and clarification. 
If you have time to run the method with one step, please let me know if you observe any difference in the performance. Technically I think it should not change much.

**Comment by bbdamodaran** [comment]
I think no need to compute the backward function two times no ?, you could also do it one single backward calculation right, as:
label_loss  = loss on labled data
unlabled_loss = loss on unlabeled data with grad rev
tot_loss  = label_loss + unlabled_loss
tot_loss.backward()
optimizer.step()

If it is correct, you can gain speed up for large networks. Please let me know whether am I correct

**Comment by ksaito-ut** [comment]
That's true.
